### ISSUE TITLE
Support for Clean Architecture Pattern using multi-module layer abstraction for domain, data & presentation modules

### ISSUE THREAD

**Issue by Rishabh-More** [dizitart/nitrite-database#202]
Well incase if you haven't been keeping up with the recent trend in the android devs community, Clean Architecture pattern is the new hot thing right now. If you've unfamiliar with clean architecture, you can read [this amazing blog post.](https://antonioleiva.com/clean-architecture-android/)

Basically clean architecture focuses on abstraction of the different layers in a code. In practice, this abstraction is achieved by having different modules in your android project (Presentation(UI), Data(Repositories) & Domain(data entities, usecases, etc.)) 
So far I can see potential in this project but the current implementation does not favour the S.O.L.I.D. principles of Clean Architecture. So for now, I'm only starring this project & will keep an eye on this one ;)

**TLDR:** Need support for multi-module project implementation. e.g. Data Entities in one module & the repository & CRUD methods implementations in another module.

**Comment by anidotnet** [comment]
Clean arch and SOLID principles are nothing new, it predates android. IMO it is more applicable for an entire application than any framework library. 

Nitrite is not an application, it is a persistence library whose one and only concern is storing data. Presentation of data is out of its scope. Moreover android is not this library's only use case. People happened to use it in server side application also (though number is less compared to android). To me, it does not make any sense to break this library into several modules based on the concerns you have mentioned. Imagine people always adding 3-4 nitrite dependencies (which could easily have been one) in their project without any real benefit.

Rather, work is currently undergoing (for v4.x) to split this library based on plugin arch and creates several moving parts so that people can add/replace default storage engine, indexing engine etc. People can use sqlite, leveldb instead of h2-mvstore, they can use lucene instead of built-in text indexing etc, which could open limitless possibilities for the community. 

My motto is, don't use any architecture principles just because it is hot and trendy in market, always use the one which brings you most benefit.
